File: dataforge-storage/src/main/java/hep/dataforge/storage/filestorage/FileTableLoader.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package hep.dataforge.storage.filestorage;

import hep.dataforge.context.Context;
import hep.dataforge.exceptions.StorageException;
import hep.dataforge.io.IOUtils;
import hep.dataforge.io.LineIterator;
import hep.dataforge.meta.Meta;
import hep.dataforge.names.Names;
import hep.dataforge.storage.api.Storage;
import hep.dataforge.storage.api.ValueIndex;
import hep.dataforge.storage.commons.DefaultIndex;
import hep.dataforge.storage.loaders.AbstractTableLoader;
import hep.dataforge.tables.MetaTableFormat;
import hep.dataforge.tables.PointParser;
import hep.dataforge.tables.SimpleParser;
import hep.dataforge.tables.TableFormat;
import hep.dataforge.values.Value;
import hep.dataforge.values.Values;
import org.apache.commons.io.FilenameUtils;
import org.jetbrains.annotations.NotNull;

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.channels.Channels;
import java.nio.file.Path;
import java.util.Iterator;
import java.util.function.Supplier;

/**
 * @author Alexander Nozik
 */
public class FileTableLoader extends AbstractTableLoader {

    public static FileTableLoader fromEnvelope(Storage storage, FileEnvelope envelope) throws Exception {
        if (FileStorageEnvelopeType.validate(envelope, POINT_LOADER_TYPE)) {
            FileTableLoader res = new FileTableLoader(storage,
                    FilenameUtils.getBaseName(envelope.getFile().getFileName().toString()),
                    envelope.meta(),
                    envelope.getFile());
            res.setReadOnly(envelope.isReadOnly());
            return res;
        } else {
            throw new StorageException("Is not a valid point loader file");
        }
    }

    private final Path path;
    //FIXME move to abstract
    private TableFormat format;
    private PointParser parser;

    /**
     * An envelope used for pushing
     */
    private FileEnvelope envelope;

    public FileTableLoader(Storage storage, String name, Meta meta, Path path) {
        super(storage, name, meta);
        this.path = path;
    }

    @Override
    public void open() throws Exception {
        if (this.meta == null) {
            this.meta = buildEnvelope(true).meta();
        }
        // read format from first line if it is not defined in meta
        if (getFormat() == null) {
            try (FileEnvelope envelope = buildEnvelope(true)) {
                new BufferedReader(Channels.newReader(envelope.getData().getChannel(), "UTF8"))
                        .lines()
                        .findFirst()
                        .ifPresent(line -> {
                            if (line.startsWith("#f")) {
                                format = MetaTableFormat.forNames(Names.of(line.substring(2).trim().split("[^\\w']+")));
                            }
                        });
            }
        }
    }

    @Override
    public void close() throws Exception {
        parser = null;
        format = null;
        if (envelope != null) {
            envelope.close();
            envelope = null;
        }
        super.close();
    }

    private FileEnvelope buildEnvelope(boolean readOnly) {
        return FileEnvelope.open(path, readOnly);
    }

    /**
     * Get writeable reusable single access envelope for this loader
     *
     * @return
     */
    private FileEnvelope getEnvelope() {
        if (this.envelope == null) {
            this.envelope = buildEnvelope(false);
        }
        return this.envelope;
    }

    @Override
    public TableFormat getFormat() {
        if (format == null) {
            if (meta().hasMeta("format")) {
                format = new MetaTableFormat(meta().getMeta("format"));
            } else if (meta().hasValue("format")) {
                format = MetaTableFormat.forNames(meta().getStringArray("format"));
            } else {
                format = null;
            }
        }
        return format;
    }

    private PointParser getParser() {
        if (parser == null) {
            parser = new SimpleParser(getFormat());
        }
        return parser;
    }

    @Override
    protected void pushPoint(Values dp) throws StorageException {
        try {
            if (!getEnvelope().hasData()) {
                getEnvelope().appendLine(IOUtils.formatCaption(getFormat()));
            }
            String str = IOUtils.formatDataPoint(getFormat(), dp);
            getEnvelope().appendLine(str);
        } catch (IOException ex) {
            throw new StorageException("Error while openning an envelope", ex);
        }
    }

    @Override
    public boolean isEmpty() {
        try {
            return (envelope != null && !envelope.hasData()) || !buildEnvelope(true).hasData();
        } catch (StorageException ex) {
            throw new RuntimeException("Can't access loader envelope", ex);
        }
    }

    @NotNull
    @Override
    public Iterator<Values> iterator() {
        try {
            FileEnvelope reader = buildEnvelope(true);
            LineIterator iterator = new LineIterator(reader.getData().getStream(), "UTF-8");
            return new Iterator<Values>() {
                @Override
                public boolean hasNext() {
                    return iterator.hasNext();
                }

                @Override
                public Values next() {
                    return transform(iterator.next());
                }
            };
        } catch (StorageException | IOException ex) {
            throw new RuntimeException(ex);
        }

    }

    private Values transform(String line) {
        return getParser().parse(line);
    }

    @Override
    public ValueIndex<Values> buildIndex(String name) {
        if (name == null || name.isEmpty()) {
            //use point number index
            return new DefaultIndex<>(this);
        } else {
            return new FilePointIndex(name, getStorage().getContext(), this::getEnvelope);
        }
    }


    private class FilePointIndex extends FileMapIndex<Values> {

        private final String valueName;

        public FilePointIndex(String valueName, Context context, Supplier<FileEnvelope> sup) {
            super(context, sup);
            this.valueName = valueName;
        }

        @Override
        protected Value getIndexedValue(Values entry) {
            return entry.getValue(valueName);
        }

        @Override
        protected String indexFileName() {
            if (getStorage().isAnonimous()) {
                return getName() + "_" + valueName;
            } else {
                return getStorage().getName() + "/" + getName() + "_" + valueName;
            }
        }

        @Override
        protected Values readEntry(String str) {
            return FileTableLoader.this.transform(str);
        }

    }
}

File: dataforge-storage/src/main/kotlin/hep/dataforge/storage/filestorage/FileTableLoader.kt
/*
 * Copyright  2018 Alexander Nozik.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package hep.dataforge.storage.filestorage

import hep.dataforge.exceptions.StorageException
import hep.dataforge.io.IOUtils
import hep.dataforge.io.LineIterator
import hep.dataforge.isAnonymous
import hep.dataforge.meta.Meta
import hep.dataforge.storage.api.Storage
import hep.dataforge.storage.api.TableLoader
import hep.dataforge.storage.api.ValueIndex
import hep.dataforge.storage.commons.DefaultIndex
import hep.dataforge.tables.*
import hep.dataforge.values.Value
import hep.dataforge.values.Values
import java.io.IOException
import java.util.*

/**
 * @author Alexander Nozik
 */
class FileTableLoader(storage: Storage, name: String, meta: Meta, file: FileEnvelope) : FileLoader(storage, name, meta, file), TableLoader {

    override val format: TableFormat by lazy {
        when {
            file.meta.hasMeta("format") -> MetaTableFormat(meta.getMeta("format"))
            file.meta.hasValue("format") -> MetaTableFormat.forNames(*meta.getStringArray("format"))
            else -> throw RuntimeException("Format definition not found")
        }
    }

    private val parser: ValuesParser by lazy {
        SimpleValuesParser(format)
    }

    override fun iterator(): MutableIterator<Values> {
        try {
            val iterator = LineIterator(file.data.stream, "UTF-8")
            return object : MutableIterator<Values> {

                override fun remove() {
                    TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
                }

                override fun hasNext(): Boolean {
                    return iterator.hasNext()
                }

                override fun next(): Values {
                    return parser.parse(iterator.next())
                }
            }
        } catch (ex: StorageException) {
            throw RuntimeException(ex)
        } catch (ex: IOException) {
            throw RuntimeException(ex)
        }

    }

    private fun buildIndex(name: String?): ValueIndex<Values> {
        return if (name == null || name.isEmpty()) {
            //use point number index
            DefaultIndex(this)
        } else {
            FilePointIndex(name)
        }
    }


    private val indexMap = HashMap<String, ValueIndex<Values>>()

    @Throws(StorageException::class)
    override fun push(dps: Collection<Values>) {
        for (dp in dps) {
            push(dp)
        }
    }

    @Synchronized
    override fun getIndex(name: String): ValueIndex<Values> {
        return indexMap.computeIfAbsent(name) { this.buildIndex(it) }
    }

    /**
     * Push point and notify all listeners
     *
     * @param dp
     * @throws StorageException
     */
    @Throws(StorageException::class)
    override fun push(dp: Values) {
        //Notifying the listener
        connectionHelper.forEachConnection(ValuesListener::class.java) {
            it.accept(dp)
        }

        try {
            val str = IOUtils.formatDataPoint(format, dp)
            file.appendLine(str)
        } catch (ex: IOException) {
            throw StorageException("Error while opening an envelope", ex)
        }
    }

//    override fun respond(message: Envelope): Envelope {
//        try {
//            if (!validator.isValid(message)) {
//                return StorageMessageUtils.exceptionResponse(message, WrongTargetException())
//            }
//            val messageMeta = message.meta
//            val operation = messageMeta.getString(ACTION_KEY)
//            when (operation) {
//                PUSH_ACTION -> {
//                    if (!messageMeta.hasMeta("data")) {
//                        //TODO реализовать бинарную передачу данных
//                        throw StorageException("No data in the push data command")
//                    }
//
//                    val data = messageMeta.getMeta("data")
//                    for (dp in ListOfPoints.buildFromMeta(data)) {
//                        this.push(dp)
//                    }
//
//                    return confirmationResponse(message)
//                }
//
//                PULL_ACTION -> {
//                    var points: List<Values> = ArrayList()
//                    when {
//                        messageMeta.hasMeta(QUERY_ELEMENT) -> points = index.query(messageMeta.getMeta(QUERY_ELEMENT)).toList()
//                        messageMeta.hasValue("value") -> {
//                            val valueName = messageMeta.getString("valueName", "")
//                            points = messageMeta.getValue("value").list.stream()
//                                    .map { `val` -> getIndex(valueName).pullOne(`val`) }
//                                    .filter { it.isPresent }.map<Values> { it.get() }
//                                    .toList()
//                        }
//                        messageMeta.hasMeta("range") -> {
//                            val valueName = messageMeta.getString("valueName", "")
//                            for (rangeAn in messageMeta.getMetaList("range")) {
//                                val from = rangeAn.getValue("from", Value.NULL)
//                                val to = rangeAn.getValue("to", Value.NULL)
//                                //                            int maxItems = rangeAn.getInt("maxItems", Integer.MAX_VALUE);
//                                points = this.getIndex(valueName).pull(from, to).toList()
//                            }
//                        }
//                    }
//
//                    val dataAn = MetaBuilder("data")
//                    for (dp in points) {
//                        dataAn.putNode(dp.toMeta())
//                    }
//                    return okResponseBase(message, true, false)
//                            .putMetaNode(dataAn)
//                            .setMetaValue("data.size", points.size)
//                            .build()
//                }
//
//                else -> throw NotDefinedException(operation)
//            }
//
//        } catch (ex: StorageException) {
//            return StorageMessageUtils.exceptionResponse(message, ex)
//        } catch (ex: UnsupportedOperationException) {
//            return StorageMessageUtils.exceptionResponse(message, ex)
//        } catch (ex: NotDefinedException) {
//            return StorageMessageUtils.exceptionResponse(message, ex)
//        }
//
//    }


    private inner class FilePointIndex(private val valueName: String) : FileMapIndex<Values>(context, file) {

        override fun getIndexedValue(entry: Values): Value {
            return entry.getValue(valueName)
        }

        override fun indexFileName(): String {
            return if (storage.isAnonymous) {
                name + "_" + valueName
            } else {
                storage.name + "/" + name + "_" + valueName
            }
        }

        override fun readEntry(str: String): Values {
            return parser.parse(str)
        }

    }
}
